### PR TITLE
Linting revamp

### PR DESCRIPTION
**Pull request by PunkPun**
This PR does quite a lot. I split everything neatly into commits.

TLDR:
* CI now fails on utility warnings
* Reduced the severity of translation errors to warnings
* Unified error formatting
* Improved a few lint tests to both be clearer and provide more info

### CODE REVIEW

File: OpenRA.Mods.Common/Lint/CheckDefaultVisibility.cs [hunk 1]
```diff
@@ -36,22 +36,22 @@ void Run(Action<string> emitError, Ruleset rules)
 				// Catch TypeDictionary errors.
 				try
 				{
-					var count = actorInfo.Value.TraitInfos<IDefaultVisibilityInfo>().Count();
+					var visibilityTypes = actorInfo.Value.TraitInfos<IDefaultVisibilityInfo>();
 
-					if (count == 0)
+					if (!visibilityTypes.Any())
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Lint/CheckDefaultVisibility.cs]: Seems like this was de-improved a bit. I suggest pulling out a variable:
`var count = visibilityTypes.Count();`
And then reverting the checks back to as it was before.

Comment (PunkPun) [reply OpenRA.Mods.Common/Lint/CheckDefaultVisibility.cs]: since I'm calling count already I suppose that makes sense

File: Makefile [hunk 2]
```diff
@@ -180,7 +194,7 @@ help:
 	@echo '  make [RUNTIME=net6] TARGETPLATFORM=unix-generic'
 	@echo
 	@echo 'to check the official mods for erroneous yaml files, run:'
-	@echo '  make [RUNTIME=net6] test'
+	@echo '  make [RUNTIME=net6] test TREAT_WARNINGS_AS_ERRORS=false'
```
Comment (abcdefg30) [inline Makefile]:
```suggestion
	@echo '  make [RUNTIME=net6] TREAT_WARNINGS_AS_ERRORS=false test'
```
to match how you call it from ci.yml?

**Review by abcdefg30 — DISMISSED** [review]

File: Makefile (hunk 2 above)
Comment (PunkPun) [reply Makefile]: good catch

**Review by abcdefg30 — DISMISSED** [review]

File: OpenRA.Mods.Common/Lint/CheckPlayers.cs [hunk 3]
```diff
@@ -92,8 +92,12 @@ void Run(Action<string> emitError, Action<string> emitWarning, MapPlayers player
 				if (playerCount > spawnPoints.Length)
 					emitError($"The map allows {playerCount} possible players, but defines only {spawnPoints.Length} spawn points.");
 
-				if (spawnPoints.Distinct().Count() != spawnPoints.Length)
-					emitError("Duplicate spawn point locations detected.");
+				var spawns = spawnPoints.ToList();
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckPlayers.cs]: This doesn't seem very efficient. What about something like:
```csharp
foreach (var spawn in spawnPoints.GroupBy(x => x))
    if (spawn.Count() > 1)
         emitError($"Duplicate spawn point location detected at ({spawn.Key}).");
```

Comment (PunkPun) [reply OpenRA.Mods.Common/Lint/CheckPlayers.cs]: It's probably of the same efficiency, but sure, it looks cleaner

File: Makefile [hunk 4]
```diff
@@ -134,6 +134,19 @@ check-scripts:
 
 test: all
 	@echo
+ifeq ($(TREAT_WARNINGS_AS_ERRORS), true)
```
Comment (pchote) [inline Makefile]: These makefile changes are really not very pretty. I suggest checking the variable directly in the utility command to avoid these.

Comment (PunkPun) [reply Makefile]: Do you know wether anything else should use this variable? Atm there's no precedence for checking environment variables outside of Program and SDL wrapper, the very lowest level of the engine

Comment (pchote) [reply Makefile]: Theres no strong precident for using flag arguments either (only the `--no-(shadow|padding)` on ConvertSpriteToPng IIRC).

Precident for using an environment variable include `MONO_LOG_LEVEL` for modifying warning/error/etc output, and Python has `PYTHONWARNINGS` that fits a similar role.

Comment (PunkPun) [reply Makefile]: I've changed it to fetch the environment variable in hopes this it's a generic enough variable to be reused

File: OpenRA.Mods.Common/Lint/CheckActorReferences.cs [hunk 5]
```diff
@@ -75,7 +75,7 @@ void CheckTrait(Action<string> emitError, ActorInfo actorInfo, TraitInfo traitIn
 
 				foreach (var requiredTrait in attribute.RequiredTraits)
 					if (!dict[v].TraitsInConstructOrder().Any(t => t.GetType() == requiredTrait || t.GetType().IsSubclassOf(requiredTrait)))
-						emitError($"Actor type {value} does not have trait {requiredTrait.Name} which is required by {traitInfo.GetType().Name}.{fieldInfo.Name}.");
+						emitError($"Actor type `{value}` does not have trait `{requiredTrait.Name}` which is required by {traitInfo.GetType().Name}.{fieldInfo.Name}.");
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckActorReferences.cs]: Why do the first two have `` but not the third?

File: OpenRA.Mods.Common/Lint/CheckTraitLocation.cs [hunk 6]
```diff
@@ -33,7 +33,7 @@ void ILintRulesPass.Run(Action<string> emitError, Action<string> emitWarning, Mo
 						var traitName = traitInfo.GetType().Name;
 						traitName = traitName.Remove(traitName.Length - 4);
 						var locations = traitLocation.SystemActors.ToString().Replace(", ", " or ");
-						emitError($"{traitName} does not belong on {actorInfo.Key}. It is a system trait meant for {locations}.");
+						emitError($"`{traitName}` does not belong on `{actorInfo.Key}`. It is a system trait meant for {locations}.");
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckTraitLocation.cs:36]: Likewise here.

File: OpenRA.Mods.Common/Lint/CheckTranslationReference.cs [hunk 7]
```diff
@@ -93,14 +93,14 @@ void ILintPass.Run(Action<string> emitError, Action<string> emitWarning, ModData
 			TestTraits(modData.DefaultRules, emitError, key =>
 			{
 				if (!translation.HasMessage(key))
-					emitError($"{key} not present in `{language}` translation.");
+					emitError($"{key} is not present in `{language}` translation.");
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckTranslationReference.cs]: No `` around `key`?

Comment (PunkPun) [reply OpenRA.Mods.Common/Lint/CheckTranslationReference.cs]: fixed

File: OpenRA.Mods.Common/Lint/CheckActorReferences.cs (hunk 5 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Lint/CheckActorReferences.cs]: I don't think we add those to multi-field fields

File: OpenRA.Mods.Common/Lint/CheckTraitLocation.cs (hunk 6 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Lint/CheckTraitLocation.cs:36]: and we definitely don't add those to multiple word fields

File: OpenRA.Mods.Common/Lint/CheckActorReferences.cs (hunk 5 above)
Comment (PunkPun) [reply OpenRA.Mods.Common/Lint/CheckActorReferences.cs]: Made the follow the convention

File: OpenRA.Mods.Common/Lint/CheckInteractable.cs [hunk 8]
```diff
@@ -42,10 +42,10 @@ void Run(Action<string> emitError, Ruleset rules, ModData modData)
 						continue;
 
 					if (HasInvalidBounds(interactable.Bounds, grid.TileSize, grid.TileScale))
-						emitError($"{nameof(interactable.Bounds)} of actor `{actorInfo.Key}` are empty or negative.");
+						emitError($"{actorInfo.Key}.{interactable.GetType().Name}.{nameof(interactable.Bounds)} are empty or negative.");
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Lint/CheckInteractable.cs:45]: Not sure if concatenating actor, trait and field all together makes anything more readable.

Comment (PunkPun) [reply OpenRA.Mods.Common/Lint/CheckInteractable.cs:45]: Providing the trait name most definitely does

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/80b1b18aafabfa31879f0d3e53772b21f5f50963...c684cfd27957230c4132a1513fa2fd45b71dbf46)

**Pull request merged**